File: backend/archive/go-backend/internal/auth/service_test.go
package auth

import (
	"context"
	"testing"
	"time"
)

type fakeProvider struct {
	authURL  string
	identity OIDCIdentity
	err      error
}

func (f fakeProvider) AuthorizationURL(state string) (string, error) {
	return f.authURL + "?state=" + state, nil
}

func (f fakeProvider) Exchange(context.Context, string) (OIDCIdentity, error) {
	return f.identity, f.err
}

type fakeStore struct {
	userID           string
	createdHash      string
	createdExpiresAt time.Time
	sessionIdentity  OIDCIdentity
	sessionOK        bool
	revokedHash      string
}

func (f *fakeStore) FindOrCreateUserByIdentity(_ context.Context, identity OIDCIdentity) (string, error) {
	f.sessionIdentity = identity
	return f.userID, nil
}

func (f *fakeStore) CreateSession(_ context.Context, userID string, tokenHash string, expiresAt time.Time) error {
	f.userID = userID
	f.createdHash = tokenHash
	f.createdExpiresAt = expiresAt
	return nil
}

func (f *fakeStore) FindIdentityBySessionHash(_ context.Context, tokenHash string, now time.Time) (OIDCIdentity, bool, error) {
	f.createdHash = tokenHash
	f.createdExpiresAt = now
	return f.sessionIdentity, f.sessionOK, nil
}

func (f *fakeStore) RevokeSession(_ context.Context, tokenHash string) error {
	f.revokedHash = tokenHash
	return nil
}

func TestServiceLoginURLReturnsProviderURLAndState(t *testing.T) {
	service := NewService(ServiceOptions{
		Provider: fakeProvider{authURL: "https://accounts.google.com/auth"},
		Store:    &fakeStore{},
		Now:      func() time.Time { return time.Unix(0, 0) },
	})

	loginURL, state, err := service.LoginURL(context.Background())
	if err != nil {
		t.Fatalf("expected login URL, got error: %v", err)
	}

	if state == "" {
		t.Fatal("expected generated state")
	}
	if loginURL != "https://accounts.google.com/auth?state="+state {
		t.Fatalf("expected state in login URL, got %q", loginURL)
	}
}

func TestServiceCompleteLoginStoresLinkedIdentityAndHashedSession(t *testing.T) {
	now := time.Date(2026, 4, 27, 12, 0, 0, 0, time.UTC)
	store := &fakeStore{userID: "user-id"}
	service := NewService(ServiceOptions{
		Provider: fakeProvider{identity: OIDCIdentity{Provider: "google", Subject: "subject", Email: "person@example.com"}},
		Store:    store,
		Now:      func() time.Time { return now },
	})

	identity, sessionToken, csrfToken, err := service.CompleteLogin(context.Background(), "code", "state")
	if err != nil {
		t.Fatalf("expected login completion, got error: %v", err)
	}

	if identity.Email != "person@example.com" {
		t.Fatalf("expected identity email, got %q", identity.Email)
	}
	if sessionToken == "" {
		t.Fatal("expected session token")
	}
	if csrfToken == "" {
		t.Fatal("expected csrf token")
	}
	if store.createdHash == "" || store.createdHash == sessionToken {
		t.Fatalf("expected hashed session token, got %q", store.createdHash)
	}
	if !store.createdExpiresAt.Equal(now.Add(DefaultSessionTTL)) {
		t.Fatalf("expected session expiry %s, got %s", now.Add(DefaultSessionTTL), store.createdExpiresAt)
	}
}

func TestServiceSessionValidatesHashedToken(t *testing.T) {
	store := &fakeStore{
		sessionIdentity: OIDCIdentity{Provider: "google", Subject: "subject", Email: "person@example.com"},
		sessionOK:       true,
	}
	service := NewService(ServiceOptions{
		Provider: fakeProvider{},
		Store:    store,
		Now:      func() time.Time { return time.Unix(0, 0) },
	})

	identity, ok := service.Session(context.Background(), "session-token")
	if !ok {
		t.Fatal("expected session to be valid")
	}
	if identity.Email != "person@example.com" {
		t.Fatalf("expected identity email, got %q", identity.Email)
	}
	if store.createdHash == "session-token" {
		t.Fatal("expected service to query by hashed token")
	}
}

func TestServiceLogoutRevokesHashedToken(t *testing.T) {
	store := &fakeStore{}
	service := NewService(ServiceOptions{
		Provider: fakeProvider{},
		Store:    store,
		Now:      func() time.Time { return time.Unix(0, 0) },
	})

	if err := service.Logout(context.Background(), "session-token"); err != nil {
		t.Fatalf("expected logout to succeed, got error: %v", err)
	}

	if store.revokedHash == "" {
		t.Fatal("expected session hash to be revoked")
	}
	if store.revokedHash == "session-token" {
		t.Fatal("expected logout to revoke hashed token")
	}
}

File: backend/src/auth/service.test.ts
import { describe, expect, test, vi } from "vitest";
import { createAuthService, normalizeUsername, type AuthStore } from "./service.js";
import type { OidcIdentity } from "./oidc.js";

describe("createAuthService", () => {
  test("exchanges OIDC identity, persists hashed session, and returns CSRF token", async () => {
    const identity: OidcIdentity = { provider: "google", subject: "subject", email: "person@example.com" };
    const provider = {
      authorizationUrl: vi.fn((state: string) => `https://provider.example/auth?state=${state}`),
      exchange: vi.fn(async () => identity)
    };
    const store: AuthStore = {
      findOrCreateUserByIdentity: vi.fn(async () => "user-1"),
      createSession: vi.fn(async () => undefined),
      findIdentityBySessionHash: vi.fn(async () => identity),
      updateProfile: vi.fn(async ({ username }) => ({ username, displayName: username })),
      revokeSession: vi.fn(async () => undefined)
    };
    const service = createAuthService({
      provider,
      store,
      now: () => new Date("2026-01-01T00:00:00Z")
    });

    const login = await service.loginUrl();
    const callback = await service.completeLogin("code", "state");
    const session = await service.session(callback.sessionToken);
    await service.logout(callback.sessionToken);

    expect(login.redirectUrl).toContain("state=");
    expect(store.findOrCreateUserByIdentity).toHaveBeenCalledWith(identity);
    expect(store.createSession).toHaveBeenCalledWith("user-1", expect.stringMatching(/^[a-f0-9]{64}$/), new Date("2026-01-31T00:00:00Z"));
    expect(callback.identity).toEqual(identity);
    expect(callback.sessionToken).not.toEqual(callback.csrfToken);
    expect(session).toEqual(identity);
    expect(store.revokeSession).toHaveBeenCalledWith(expect.stringMatching(/^[a-f0-9]{64}$/));
  });

  test("normalizes and validates usernames before updating profile", async () => {
    const provider = {
      authorizationUrl: vi.fn(),
      exchange: vi.fn()
    };
    const store: AuthStore = {
      findOrCreateUserByIdentity: vi.fn(),
      createSession: vi.fn(),
      findIdentityBySessionHash: vi.fn(),
      updateProfile: vi.fn(async ({ username }) => ({ username, displayName: username })),
      revokeSession: vi.fn()
    };
    const service = createAuthService({ provider, store });

    await expect(service.updateProfile({ userId: "user-1", username: "  June_Park  " })).resolves.toEqual({
      username: "June_Park",
      displayName: "June_Park"
    });
    expect(store.updateProfile).toHaveBeenCalledWith({
      userId: "user-1",
      username: "June_Park",
      displayName: "June_Park"
    });
    expect(() => normalizeUsername("no")).toThrow("username must be 3-24 characters using letters, numbers, or underscores");
  });
});
